From 41496085c830f47c8213221429703e93bd1a4e9c Mon Sep 17 00:00:00 2001
From: Xianzhu Wang <wangxianzhu@chromium.org>
Date: Wed, 22 Feb 2023 01:06:34 +0000
Subject: [PATCH] Ensure safe access to children during ui::Layer::OnDeviceScaleFactorChanged()

(cherry picked from commit d16c4344b0c9881b5b00a5234b6b5b0619039b6f)

Bug: 1413539
Change-Id: Ia63a3177bb53019afbff0fa2c1b8b2da57c7246f
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4265793
Auto-Submit: Xianzhu Wang <wangxianzhu@chromium.org>
Commit-Queue: Scott Violet <sky@chromium.org>
Reviewed-by: Scott Violet <sky@chromium.org>
Reviewed-by: Robert Flack <flackr@chromium.org>
Commit-Queue: Robert Flack <flackr@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1107099}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4277707
Commit-Queue: Xianzhu Wang <wangxianzhu@chromium.org>
Cr-Commit-Position: refs/branch-heads/5481@{#1246}
Cr-Branched-From: 130f3e4d850f4bc7387cfb8d08aa993d288a67a9-refs/heads/main@{#1084008}
---

--- a/ui/compositor/layer.cc
+++ b/ui/compositor/layer.cc
@@ -1338,7 +1338,16 @@
     delegate_->OnDeviceScaleFactorChanged(old_device_scale_factor,
                                           device_scale_factor);
   }
+
+  // We may add or remove children during child->OnDeviceScaleFactorChanged().
+  std::vector<base::WeakPtr<Layer>> weak_children(children_.size());
   for (auto* child : children_) {
+    weak_children.push_back(child->weak_ptr_factory_.GetWeakPtr());
+  }
+  for (auto& child : weak_children) {
+    if (!child) {
+      continue;
+    }
     child->OnDeviceScaleFactorChanged(device_scale_factor);
 
     // A child layer may have triggered a delegate or an observer to delete
